More flexible KCard click event handling

MisRob opened the pull request:
## Description

`KCard` is updated for more flexible click event handling. Previously, its title text was always `router-link`, `to` prop was required, and the card would navigate to the target on click. The new behavior is:

- `click` event is always emitted on card click
- `to` prop is not required
- when `to` is provided, the title text is `router-link`, and the card navigates to the target
- when `to` is not provided, the title text is `span`, and the card doesn't navigate anywhere

Contains related a11y changes that increase screen reader reliability overall:

- Make title prop required and allow title customization via a scoped slot
- Ensure reliable screen readers announcements no matter whether card is link or no, and no matter whether the title is customized via the title slot.

#### Issue addressed

A new use-case in Studio that @akolson works on in https://github.com/learningequality/studio/pull/4803 where a side panel needs to be toggled on card click without navigating to a new URL.

![Screenshot from 2024-11-13 16-33-50](https://github.com/user-attachments/assets/895c7406-b46d-4c59-afd2-1a0404095c8c)

## Changelog
<!-- [DO NOT REMOVE-USED BY GH ACTION] CHANGELOG START -->

<!--
  - Fill in the changelog item(s) below. If there are more groups of closely
    related changes, prepare more changelog items for each one of them.
    At a minimum, always separete non-breaking changes from breaking changes.
  - This needs to be pasted to CHANGELOG.md before merging a PR.
  - See changelog guidelines https://www.notion.so/learningequality/DRAFT-Changelog-Guidelines-106b6ebbdeda4ba5b3b3e7c490c5a4fe and existing
    items in CHANGELOG.md as examples
 -->

  - **Description:** Make the `title` prop required
  - **Products impact:** updated API
  - **Addresses:** A new use-case in Studio in https://github.com/learningequality/studio/pull/4803 where a side panel needs to be toggled on card click without changing URL.
  - **Components:** KCard
  - **Breaking:** yes
  - **Impacts a11y:** no
  - **Guidance:** Even if you use the `title` slot, pass the title text via the `title` prop.

  - **Description:** Change the `title` slot into a scoped slot
  - **Products impact:** updated API
  - **Addresses:** A new use-case in Studio in https://github.com/learningequality/studio/pull/4803 where a side panel needs to be toggled on card click without changing URL.
  - **Components:** KCard
  - **Breaking:** no
  - **Impacts a11y:** no
  - **Guidance:**  Consider using the slot's `textTitle` attribute to achieve more intuitive code when customizing the title.

  - **Description:** Emit `click` event when card is clicked.
  - **Products impact:** updated API
  - **Addresses:** A new use-case in Studio in https://github.com/learningequality/studio/pull/4803 where a side panel needs to be toggled on card click without changing URL.
  - **Components:** KCard
  - **Breaking:** no
  - **Impacts a11y:** no
  - **Guidance:** -

  - **Description:** Make `to` prop optional and when not provided, do not render the title text as `router-link` but rather as `span`.
  - **Products impact:** updated API
  - **Addresses:** A new use-case in Studio in https://github.com/learningequality/studio/pull/4803 where a side panel needs to be toggled on card click without changing URL.
  - **Components:** KCard
  - **Breaking:** no
  - **Impacts a11y:** no
  - **Guidance:** -

  - **Description:** Ensure reliable screen readers announcements no matter whether card is link or no, and no matter whether the title is customized via the title slot
  - **Products impact:** bugfix
  - **Addresses:** A new use-case in Studio in https://github.com/learningequality/studio/pull/4803 where a side panel needs to be toggled on card click without changing URL.
  - **Components:** KCard
  - **Breaking:** no
  - **Impacts a11y:** yes
  - **Guidance:** -

<!-- [DO NOT REMOVE-USED BY GH ACTION] CHANGELOG END -->

## Steps to test

- You could run the devserver, play around with `to` and `@click` on the [DocsKCard](https://github.com/learningequality/kolibri-design-system/blob/develop/docs/pages-components/DocsKCard.vue#L3-L11) and preview live example on `KCard` docs page.
- Is [the updated documentation section](https://deploy-preview-825--kolibri-design-system.netlify.app/kcard/#click-navigation) clear?
- Can you see any functional, markup, or a11y regressions?

## Testing checklist
<!-- Complete the checklist before submitting a PR; delete anything that doesn't apply -->

- [x] Contributor has fully tested the PR manually
- [ ] ~If there are any front-end changes, before/after screenshots are included~
- [x] Critical and brittle code paths are covered by unit tests
- [x] The change is described in the changelog section above

## Reviewer guidance
<!-- Delete anything that doesn't apply so your reviewer knows what to check for -->

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

## Comments
<!-- Any additional notes you'd like to add -->

MisRob committed:
Make navigation on card click optional

and add a click event. This allows for
a custom handler.

AllanOXDi reviewed (approved):
Looks good to me !Thanks Misha, maybe @akolson and @AlexVelezLl has something to say

AlexVelezLl commented on lib/cards/KCard.vue:
```diff
@@ -54,6 +55,40 @@
+          tabindex="0"
+          data-focus="true"
+          class="title"
+          :aria-label="title"
```
Here `title` isnt allways necessarily defined, if we use `$slots.title` we wont get the announcement of any title, because this aria-label will be empty, and the slot will be surrounded by an `aria-hidden`. I think we should do an special management here and get the text of the slot to add this as aria-label when the title is being sent as a slot rather than as a prop.

MisRob replied:
Thanks @AlexVelezLl, that's a good catch - I completely forgot there's two title pathways.

> get the text of the slot to add this as aria-label when the title is being sent as a slot rather than as a prop

I thought and played around it and it seems to be tricky. (1) I don't see a straightforward way to obtain it. E.g. for

```
<template #title>
  <span>title</span>
</template>
```

`this.$slots.title`  is an array of `VNode`s and I can't see any attribute on `VNode` that would give the text, neither I've been able to access HTML element. Perhaps you can notice something I missed?

(2) However in any case, any HTML content can be passed via the slot, and I think we can't be 100% assuming it would be suitable `aria-label`.

So I wonder if we should just add `ariaLabel` required prop? For the sake of simplicity and explicitness, we may consider requiring it even when the `title` is passed via a prop.

MisRob replied:
Even though maybe requiring it when `title` prop is there is way too strict and would be an unnecessary hassle. Technically, it's not a problem to make customized check for only when the title slot is used. Will wait to hear from you, perhaps you will have opinion.

AlexVelezLl replied:
In KOptionalText we get the slots text in this way: https://github.com/learningequality/kolibri-design-system/blob/39ad629331d74cf90622e86b3a6012b384718fd8/lib/KOptionalText.vue#L54-L74, I think we can combine these approaches, and have an ariaLabel prop, and have this `getSlotsText` as a fallback if we have a slots.title and dont have an aria-label. So for example if we have complex title slots we can use the ariaLabel prop to specifically set the aria-label as needed. But for the most basic use cases like

```
<template #title>
  <span>title</span>
</template>
```

We can live just with the `getSlotsText`.

MisRob replied:
@AlexVelezLl resolved in a way we agreed on Slack

AlexVelezLl commented on lib/cards/KCard.vue:
```diff
@@ -436,23 +472,26 @@
+        /**
+         * Emitted when a card is clicked
+         */
+        this.$emit('click');
```
Its usually useful to emit the event object here too. I have found myself many times updating some KDS components because I need the event object, so I think it doent hurt if we emit it too since the beginning.

MisRob replied:
Ah definitely. Done.

AlexVelezLl reviewed (changes requested):
Hi @MisRob! I just noted a couple of things I would like to discuss about :)

MisRob committed:
Add event payload

MisRob committed:
Make title prop required

and allow title customization
via a scoped slot.

Contains related work to ensure reliable
screen readers announcements
no matter whether card is link or no,
and no matter whether the title is
customized via the title slot.

MisRob commented:
Thanks @AllanOXDi @AlexVelezLl, all feedback is resolved, changelog and PR description updated with new information. I tested screenreaders in all modes. As soon as we have some working features in Kolibri, we will loop in Radina for final confirmation (outside this PR though I'd suggest because there's way more a11y we'll be testing in the final products)

MisRob commented:
In the new docs example, please ignore the thumbnail overflow issue - I shoud have this fixed in another branch

AlexVelezLl reviewed (approved):
Thank you @MisRob! Looks good to me!
